includes/class-llms-rest-authentication.php

```diff
@@ -17,13 +17,271 @@
+		$consumer_key = llms_rest_api_hash( $consumer_key );
+
+		$key_id = $wpdb->get_var( $wpdb->prepare( "SELECT id FROM {$wpdb->prefix}lifterlms_api_keys WHERE consumer_key = %s", $consumer_key ) );
+
+		if ( $key_id ) {
+			return LLMS_REST_API()->keys()->get( $key_id );
+		}
+
+		return false;
+
+	}
+
+	/**
+	 * Locate credentials in the $_SERVER superglobal.
+	 *
+	 * @since [version]
+	 *
+	 * @param string $key_var Variable name for the consumer key.
+	 * @param string $secret_var Variable name for the consumer secret.
+	 * @return array|false
+	 */
+	private function get_credentials( $key_var, $secret_var ) {
+
+		// Use `filter_var()` instead of `llms_filter_input()` due to PHP bug with `filter_input()`: https://bugs.php.net/bug.php?id=44779.
+		$key    = isset( $_SERVER[ $key_var ] ) ? filter_var( wp_unslash( $_SERVER[ $key_var ] ), FILTER_SANITIZE_STRING ) : null;
+		$secret = isset( $_SERVER[ $secret_var ] ) ? filter_var( wp_unslash( $_SERVER[ $secret_var ] ), FILTER_SANITIZE_STRING ) : null;
+
+		if ( ! $key || ! $secret ) {
+			return false;
+		}
+
+		return compact( 'key', 'secret' );
+
+	}
+
+	/**
+	 * Retrieve the auth error object.
+	 *
+	 * @since [version]
+	 *
+	 * @return WP_Error|null
+	 */
+	protected function get_error() {
+		return $this->error;
+	}
+
+	/**
+	 * Determine if the request is a request to a LifterLMS REST API endpoint.
+	 *
+	 * @since [version]
+	 *
+	 * @return bool
+	 */
+	protected function is_rest_request() {
+
+		$request = llms_filter_input( INPUT_SERVER, 'REQUEST_URI', FILTER_SANITIZE_STRING );
+		if ( empty( $request ) ) {
+			return false;
+		}
+
+		$request = esc_url_raw( wp_unslash( $request ) );
+		$prefix  = trailingslashit( rest_get_url_prefix() );
+
+		$core = ( false !== strpos( $request, $prefix . 'llms/' ) );
+
+		// Allow 3rd parties to use core auth.
+		$external = ( false !== strpos( $request, $prefix . 'llms-' ) );
+
+		return apply_filters( 'llms_is_rest_request', $core || $external, $request );
+
+	}
+
+	/**
+	 * Get api credentials from headers and then basic auth.
+	 *
+	 * @since [version]
+	 *
+	 * @return array|false
+	 */
+	protected function locate_credentials() {
+
+		// Attempt to get creds from headers.
+		$creds = $this->get_credentials( 'HTTP_X_LLMS_CONSUMER_KEY', 'HTTP_X_LLMS_CONSUMER_SECRET' );
+		if ( $creds ) {
+			return $creds;
+		}
+
+		// Attempt to get creds from basic auth.
+		$creds = $this->get_credentials( 'PHP_AUTH_USER', 'PHP_AUTH_PW' );
+		if ( $creds ) {
+			return $creds;
+		}
+
+		return false;
+
+	}
+
+	/**
+	 * Return a WWW-Authenticate header error message when incorrect creds are supplied
+	 *
+	 * @link https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/WWW-Authenticate
+	 *
+	 * @param WP_REST_Response $response Current response being served.
+	 * @return WP_REST_Response
+	 */
+	public function send_unauthorized_headers( $response ) {
+
+		if ( is_wp_error( $this->get_error() ) ) {
+			$auth_message = __( 'LifterLMS REST API', 'lifterlms' );
```

Comment:
Maybe we want to send a more meaningful error message

Reply:
Check out the link there (https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/WWW-Authenticate).

The WooCommerce rest api does add an error here however from my understanding of the header's intention it's supposed to just describe the area. The error message itself (which would be included in the rest body) would be the error message.

This is just an additional header.

Semantically I think the current message here is correct and boy do I love me some semantics :-p

Reply:
Yes Woo misled me.
I think you're right and I should've checked the @link before questioning.
My bad.